frontend/src/queries/nodes/DataVisualization/displayLogic.ts

```diff
@@ -0,0 +1,89 @@
+import * as d3 from 'd3'
+import { actions, afterMount, connect, kea, key, path, props, reducers } from 'kea'
+import { subscriptions } from 'kea-subscriptions'
+
+import { GoalLine } from '~/queries/schema'
+
+import { dataVisualizationLogic } from './dataVisualizationLogic'
+import type { displayLogicType } from './displayLogicType'
+
+export interface DisplayLogicProps {
+    key: string
+}
+
+export const displayLogic = kea<displayLogicType>([
+    key((props) => props.key),
+    path(['queries', 'nodes', 'DataVisualization', 'displayLogic']),
+    props({ key: '' } as DisplayLogicProps),
+    connect({
+        values: [dataVisualizationLogic, ['yData', 'query']],
+        actions: [dataVisualizationLogic, ['setQuery']],
+    }),
+    actions(({ values }) => ({
+        addGoalLine: () => ({ yData: values.yData }),
+        updateGoalLine: (goalLineIndex: number, key: string, value: string | number) => ({
+            goalLineIndex,
+            key,
+            value,
+        }),
+        removeGoalLine: (goalLineIndex: number) => ({ goalLineIndex }),
+        setGoalLines: (goalLines: GoalLine[]) => ({ goalLines }),
+    })),
+    reducers({
+        goalLines: [
+            [] as GoalLine[],
+            {
+                addGoalLine: (prev, { yData }) => {
+                    const yDataFlat = yData?.flatMap((n) => n.data) ?? []
+                    const yDataAvg = Math.round(d3.mean(yDataFlat) ?? 0)
+
+                    prev.push({
+                        label: 'Q4 Goal',
+                        value: yDataAvg ?? 0,
+                    })
+                    return [...prev]
```

Comment:
You shouldn't modify existing data in reducers, but always create new objects. Otherwise there are subtle bugs.

```suggestion
                    return [...prev, {
                        label: 'Q4 Goal',
                        value: yDataAvg ?? 0,
                    }]
```